### PR TITLE
Use VC140_ON_VS2017 to decide on whether to use VS 2017 to compile for vc=14.0

### PR DESCRIPTION
**Pull request by isuruf**
cc @msarahan, @mingwandroid

### CODE REVIEW

**Comment by mariusvniekerk** [comment]
We should add support for VS2019 as well

**Comment by msarahan** [comment]
OK, but why?  Why not just use VS2017 with 14.1 toolchain?

**Comment by isuruf** [comment]
We are using VS2015 on appveyor (VS2015 image has VS2008 installed, but not VS2017, so it's easier) and we want to use the same thing on azure until azure works reliably to switch.

**Comment by isuruf** [comment]
Is VS2019 released yet? If not, let's push it off for now. (If you have time, send a PR to my fork).

File: vs2015/install_activate.bat [hunk 1]
```diff
@@ -4,6 +4,18 @@ set VER=14
 mkdir "%PREFIX%\etc\conda\activate.d"
 COPY "%RECIPE_DIR%\activate.bat" "%PREFIX%\etc\conda\activate.d\vs%YEAR%_compiler_vars.bat"
 
+echo if "%%VC140_ON_VS2017%%" == "" ( >> "%PREFIX%\etc\conda\activate.d\vs%YEAR%_compiler_vars.bat"
+
+IF "%cross_compiler_target_platform%" == "win-64" (
+  echo SET "CMAKE_GENERATOR=Visual Studio 15 2017 Win64" -T "v140" >> "%PREFIX%\etc\conda\activate.d\vs%YEAR%_compiler_vars.bat"
```
Comment (msarahan) [inline vs2015/install_activate.bat]: I'm not sure about the quoting here.  I'm not sure that the -T "V140" will get included properly.  I think the proper quoting *might* be:

SET "CMAKE_GENERATOR=""Visual Studio 15 2017 Win64"" -T v140"

Comment (isuruf) [reply vs2015/install_activate.bat]: We use `cmake -G "%CMAKE_GENERATOR"` and with my quoting this turns into
`cmake -G "Visual Studio 15 2017 Win64" -T "v140"`

with yours, it's
`cmake -G """Visual Studio 15 2017 Win64"" -T v140"`

Comment (msarahan) [reply vs2015/install_activate.bat]: Cool, thanks for checking.

**Comment by mariusvniekerk** [comment]
@msarahan anything left to do for this one?

**Comment by msarahan** [comment]
I think this needs better documentation.  Where is ``VS2017_VCVARSALL`` supposed to come from?  How is this intended to be used?  Why is it here?  Keep in mind what you're asking here is not just for this to be available to you in your packages, it's also to everyone else.  Is it going to potentially be work for us to figure out/explain to other people later?  I think right now the answer is "yes" but I would like it to be "no"

The alternative to having this here is to have these activation scripts in conda-forge packages instead.  In the past, we've tried to avoid that so that we avoid divergence in compiler flags and the associated incompatibility.  Is it worth exploring an override scheme that allows conda-forge to customize and add to compiler activation stuff as you are doing here?  Maybe it's a package that depends on vs2015_win-64, for example, but adds in an extra activate.d script that is alphabetically after the vs2015_win-64 one, and which overrides the stuff like you do here.  You'd need to copy in the run_exports, since those are not transitive.  You'd set c_compiler and friends in conda_build_config.yaml to use this wrapper/extender package instead.

Does that seem reasonable?

**Comment by isuruf** [comment]
>  Where is VS2017_VCVARSALL supposed to come from? How is this intended to be used? Why is it here? 

From `conda_build_config.yaml`. We are going to set the environment variable in our conda_build_config.yaml

**Comment by isuruf** [comment]
> Maybe it's a package that depends on vs2015_win-64

That's not going to work. If vs2015_win-64 calls vcvarsall.bat (which doesn't work properly with vc14.0 on vs2017) then subsequent calls to vcvarsall.bat doesn't work.

**Comment by msarahan** [comment]
OK, please document this with comments in the code itself.  As I said before, I don't want Anaconda to spend time answering any questions on this.  If you want it in our package, I need it to be self-describing.  It is not sufficient for there to only be the discussion in this PR.  It needs to be in the code itself.

**Comment by msarahan** [comment]
I pushed more text.  Please remember the context Anaconda is in.  It's more than just conda-forge.  Anything you want to add to our world, ESPECIALLY weird, non-standard things to compensate for awkward CI peculiarities, needs a lot of description and explanation.

Everything needs to be explained in terms of not just what it does, but why, and ideally for how long will such weirdness be necessary (that's also an important part of the "why").

We can't make these additions for you every time.  Please include more description going forward.

**Comment by msarahan** [comment]
Packages are up on the c3i_test2 channel.  Please test them.

**Pull request merged**